### PR TITLE
Cctools and ld64

### PR DESCRIPTION
**Pull request by isuruf**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your
      recipe (see
      [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos)
      for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by isuruf** [comment]
@mingwandroid, do you know if `libxar.1.dylib` is available on all osx versions?

File: recipes/cctools-ld64/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      home: https://opensource.apple.com/source
+      license: Apple Public Source License 2.0
+      license_family: Other
+      license_file: cctools/ld64/APPLE_LICENSE
+      summary: Darwin Mach-O linker
+
+extra:
+  recipe-maintainers:
+    - isuruf
+    - mingwandroid
```
Comment (isuruf) [inline recipes/cctools-ld64/meta.yaml]: @mingwandroid, can you confirm that you want to be a maintainer?

**Comment by isuruf** [comment]
@conda-forge/staged-recipes, ready for a review.

File: recipes/cctools-ld64/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        - libcxx  # [osx]
+      run:
+        - {{ pin_compatible("tapi") }}
+        - libcxx  # [osx]
+    test:
+      commands:
+        - test -f $PREFIX/bin/{{ macos_machine }}-ld
+    about:
+      home: https://opensource.apple.com/source
+      license: Apple Public Source License 2.0
```
Comment (chrisburr) [inline recipes/cctools-ld64/meta.yaml]:
```suggestion
      license: APSL-2.0
```

File: recipes/cctools-ld64/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  - name: cctools
+    version: {{ cctools_version }}
+    script: install-cctools.sh
+    test:
+      commands:
+        - test -f $PREFIX/libexec/as/x86_64/as
+        - test -f $PREFIX/bin/{{ macos_machine }}-as
+    about:
+      home: https://opensource.apple.com/source
+      license: Apple Public Source License 2.0
```
Comment (chrisburr) [inline recipes/cctools-ld64/meta.yaml]: SPDX identifier For consistency with other recipes
```suggestion
      license: APSL-2.0
```

**Comment by isuruf** [comment]
@chrisburr, ready for a final review

**Comment by isuruf** [comment]
Thanks for the review

**Comment by mingwandroid** [comment]
You've now rebuilt brand new macOS binary tools (with TAPI support?) and linker with our own slightly older and smelly binary tools and linkers? If I am correct about this then this is a milestone (I was nervous about it in all honesty). You rock @isuruf.

But once you build these it'd be cool to build them again, i.e. new with new, flush out the old smell!

This is really great to see. We can now think about an OpenMacOSSDK that we bundle I think. We can also cross compile MacOS software on Linux without too much extra work.

@isuruf, I doff my hat to you once more.

**Comment by isuruf** [comment]
> You've now rebuilt brand new macOS binary tools (with TAPI support?) and linker with our own slightly older and smelly binary tools and linkers?

Yes. I checked that I can link with MacOS 10.14 TBDs

> But once you build these it'd be cool to build them again, i.e. new with new, flush out the old smell!

Yes, on my todo list. I need to first create a new `clang_bootstrap_osx-64` as well. Then I can remove the installation of the the conda compiler in the build script and use `clang_bootstrap_osx-64`. Issue was that I was using clang 8.0.1 with the linker from 4.0.1 and it didn't work. I had to use 4.0.1 since there was no `clang_bootstrap_osx-64=4.0.1`.

**Comment by mingwandroid** [comment]
I wonder how automatable you feel this stuff is. Once I think too much about it my head hurts!

**Comment by isuruf** [comment]
All you need is a `clang_bootstrap_osx-64` that uses system compilers to bootstrap the conda compilers and then create a new `clang_bootstrap_osx-64` using the conda compilers and build again.

**Comment by mingwandroid** [comment]
So, should we plan to work on an openMacOSSDK package @conda-forge/core? I know there were licensing things to be discussed and resolved there, but I think it is worth persuing.

File: recipes/cctools-ld64/meta.yaml (hunk 1 above)
Comment (mingwandroid) [reply recipes/cctools-ld64/meta.yaml]: Yes, I would like that. Thank you. If I'm added next time this is updated that's also fine by me.

**Comment by isuruf** [comment]
@mingwandroid, can you open an issue in https://github.com/conda-forge/conda-forge.github.io ? We can discuss there

**Comment by mingwandroid** [comment]
> @mingwandroid, do you know if `libxar.1.dylib` is available on all osx versions?

I think it is yes. We can provide it easily though if you prefer that?

**Comment by isuruf** [comment]
OSX is shipping 1.8.dev and ld64 needs 1.8.dev if I remember correctly. I couldn't find sources for 1.7 let alone 1.8.dev.

**Comment by mingwandroid** [comment]
Oh Apple!

**Comment by isuruf** [comment]
Found the sources for xar at https://opensource.apple.com/tarballs/xar/

**Comment by mingwandroid** [comment]
Some recipes I used back in the day!

https://github.com/AnacondaRecipes/aggregate/search?q=xar&unscoped_q=xar

Github's direct browse didn't work so I searched it.

I had a bootstrap xar, roughly synonymous with static, but also taking care to make it fully static so it bundles any deps. This is what we are now doing for conda-package-handling with libarchive_static FWIW.

**Comment by mingwandroid** [comment]
BTW mainitaining xar support (and extending it into ELF if we can) would be another string to our bow.

It allows LLVM bitcode to be emitted instead of (and as well as) machine code. Apple require that for App store submissions only bitcode is presented (at least for iPhone, but the reason is to give some arch independence to Apple should they wish to jump to PowerPC or something without having to throw away the App Store and rebuild it) ..

**Comment by mingwandroid** [comment]
Note I used a more modern fork of xar instead of Apple's

https://github.com/mackyle/xar/archive/xar-{{ version }}.tar.gz

**Comment by isuruf** [comment]
Apple seems to be maintaining it better than https://github.com/mackyle/xar. Saw somewhere that Apple's fork is the only one with some CVE fixes

**Comment by mingwandroid** [comment]
Great! We definitely want our build toolchain to be rock solid.

**Comment by mingwandroid** [comment]
I'll poke at linux cross packages too if people would be intersted in that (and I find a few hours behind the sofa somewhere).

**Pull request merged**